logstash-core/lib/logstash/instrument/metric.rb

```diff
@@ -4,36 +4,38 @@
 require "concurrent"
 
 module LogStash module Instrument
-  class MetricNoKeyProvided < Exception; end
-  class MetricNoBlockProvided < Exception; end
+  class MetricException < Exception; end
+  class MetricNoKeyProvided < MetricException; end
+  class MetricNoBlockProvided < MetricException; end
+  class MetricNoNamespaceProvided < MetricException; end
 
   class Metric
-    attr_reader :collector, :base_key
+    attr_reader :collector, :namespace_information
 
-    def initialize(collector, base_key = nil)
+    public
+    def initialize(collector, namespace = nil)
       @collector = collector
-      @base_key = base_key
+      @namespace_information = Array(namespace).map(&:to_sym)
     end
 
     def increment(key, value = 1)
-      collector.push(:counter, merge_keys(key), :increment, Concurrent.monotonic_time, value)
+      validate_key!(key)
+      collector.push(namespace_information, key, :counter, :increment, value)
     end
 
     def decrement(key, value = 1)
-      collector.push(:counter, merge_keys(key), :decrement, Concurrent.monotonic_time, value)
+      validate_key!(key)
+      collector.push(namespace_information, key, :counter, :decrement, value)
     end
 
     # might be worth to create a block interface for time based gauge
     def gauge(key, value)
-      collector.push(:gauge, merge_keys(key), Concurrent.monotonic_time, value)
+      validate_key!(key)
+      collector.push(namespace_information, key, :gauge, value)
     end
 
-    def namespace(key)
-      Metric.new(collector, merge_keys(key.to_sym))
-    end
-
-    # I think this should have his own values.
     def time(key, &block)
+      validate_key!(key)
       if block_given?
         start_time = Concurrent.monotonic_time
```

Comment:
yes + 1, I did the test yesterday after pushing the commit and didnt revisit it.